[caffe2] Remove OperatorBase::newstyle_outputs_

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #67093

`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

peterbell10 committed:
[caffe2] Remove Operator::newstyle_outputs_

`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove Operator::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove Operator::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove Operator::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

dagitses commented on caffe2/core/export_caffe2_op_to_c10.h:
```diff
@@ -79,7 +80,13 @@ inline void _call_caffe2_op_from_c10(
+    outputs_c2[i] = caffe2::Tensor(outputs.get(i));
+  }
+
+  outputs_c2 = (*call_op)(schema, std::move(inputs), std::move(outputs_c2));
```
is it guaranteed that `outputs_c2`'s size afterward is no greater than its size before?

Should we `assert(num_outputs <= outputs_c2.size());`?

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 commented:
@dagitses PTAL

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

dagitses commented on caffe2/core/export_caffe2_op_to_c10.h:
```diff
@@ -74,13 +75,21 @@ inline void _call_caffe2_op_from_c10(
   } else {
     AT_ASSERT(preallocated_outputs.isTensorList());
     outputs = std::move(preallocated_outputs).toTensorList();
+    TORCH_INTERNAL_ASSERT(num_outputs == outputs.size());
```
this assert is triggering with internal tests. Looking for a repro unless you know why.

peterbell10 replied:
I've changed it to assign `num_outputs = outputs.size()` to avoid any problems. I think it must be happening when the operator returns a `TensorList` so `num_outputs` isn't encoded in the schema.

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 commented:
@dagitses CI failures are unrelated and happening on master as well.

peterbell10 committed:
Rebase after master CI break on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 commented:
@dagitses ping

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

peterbell10 commented:
@dagitses ping

peterbell10 committed:
Only build IValueInterface when necessary on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

peterbell10 commented:
@dagitses ping

peterbell10 committed:
Rebase and fix merge conflicts on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 commented:
@dagitses ping

peterbell10 committed:
Rebase on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge this please

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

peterbell10 commented:
@dagitses is there any interest in landing this?

peterbell10 committed:
Rebase on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

kit1980 commented:
@dagitses do you think this is ok to merge?

peterbell10 commented:
@kit1980 is it possible to take over landing this stack?

malfet reviewed (approved):
LGTM, I think we don't need internal testing for this one

malfet commented:
@pytorchbot rebase

peterbell10 committed:
Update on "[caffe2] Remove OperatorBase::newstyle_outputs_"


`OperatorBase` maintains `output_tensors_` and `newstyle_outputs_`
which hold the same list of tensors except one is
`vector<caffe2::Tensor>` and the other is `List<at::Tensor>`.

This instead maintains only `output_tensors_` and handles the
conversions inside of export_caffe2_op_to_c10.

Differential Revision: [D32289811](https://our.internmc.facebook.com/intern/diff/D32289811)

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

kit1980 commented:
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289811).

kit1980 commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot merge -f "caffe2 only change"
